src/nxt_cgroup.c

```diff
@@ -0,0 +1,158 @@
+
+
+static char *nxt_mk_cgpath(const char *dir);
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    int   ret, len;
+    FILE  *fp;
+    char  *cgpath;
+    char  cgprocs[PATH_MAX - 13];   /* - 13 for '/cgroup.procs' */
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.path == NULL)
+    {
+        return NXT_OK;
+    }
+
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgpath == NULL) {
+        return NXT_ERROR;
+    }
+    nxt_fs_mkdir_all((const u_char *) cgpath, 0777);
+
+    snprintf(cgprocs, sizeof(cgprocs), "%s/cgroup.procs", cgpath);
+    fp = fopen(cgprocs, "we");
+    if (fp == NULL) {
+        ret = NXT_ERROR;
+        goto out_free;
+    }
+
+    ret = NXT_OK;
+
+    setvbuf(fp, NULL, _IONBF, 0);
+    len = fprintf(fp, "%d\n", process->pid);
+    if (len < 0) {
+        ret = NXT_ERROR;
+    }
+
+    fclose(fp);
+
+out_free:
+
+    nxt_free(cgpath);
+
+    return ret;
+}
+
+
+void
+nxt_cgroup_cleanup(const nxt_process_t *process)
+{
+    char  *cgroot, *cgpath, *ptr;
+
+    cgroot = nxt_mk_cgpath("");
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgroot == NULL || cgpath == NULL) {
+        goto out_free;
+    }
+
+    while (*cgpath != '\0' && strcmp(cgroot, cgpath) != 0) {
+        rmdir(cgpath);
+        ptr = strrchr(cgpath, '/');
+        *ptr = '\0';
+    }
+
+out_free:
+
+    nxt_free(cgroot);
+    nxt_free(cgpath);
+}
+
+
+static char *
+nxt_mk_cgpath(const char *dir)
+{
+    char  *path, *buf, *ptr;
+
+    path = buf = ptr = NULL;
+
+    /*
+     * If the path from the config is relative, we need to make
+     * the cgroup path include the main unit processes cgroup. I.e
+     *
+     *   NXT_CGROUP_ROOT/<main process cgroup>/<cgroup path>
+     */
+    if (dir[0] != '/') {
+        int      i;
+        bool     found;
+        FILE     *fp;
+        size_t   len;
+        ssize_t  nread;
+
+        fp = fopen("/proc/self/cgroup", "re");
+        if (fp == NULL) {
+            return NULL;
+        }
+
+        found = false;
+        while ((nread = getline(&buf, &len, fp)) != -1) {
+            if (strncmp(buf, "0::", 3) != 0) {
+                continue;
+            }
+
+            found = true;
+            break;
+        }
+
+        fclose(fp);
+
+        if (found == false) {
+            goto out_free_buf;
+        }
+
+        buf[nread - 1] = '\0';  /* lose the trailing '\n' */
+        ptr = buf;
+        for (i = 0; i < 2; i++) {
+            ptr = strchr(ptr, ':');
+            if (ptr == NULL) {
+                goto out_free_buf;
+            }
+
+            ptr++;
+        }
+    }
+
+    path = nxt_malloc(PATH_MAX);
+    if (path == NULL) {
+        goto out_free_buf;
+    }
+
+    snprintf(path, PATH_MAX, NXT_CGROUP_ROOT "%s%s%s",
+             (ptr != NULL) ? ptr : "", (ptr != NULL) ? "/" : "", dir);
```

Comment:
Should we check for truncation here (and in a few other `snprintf(3)` calls)?

Reply:
Not sure it's worth it. We also do a sanity check in the config, though that could be enhanced slightly to include space for config.procs.

Reply:
Can you point to the checks (file + line) that would make this redundant?